Source/Core/LRUCache.js

```diff
@@ -0,0 +1,174 @@
+define([
+    './defined',
+    './defineProperties',
+    './getTimestamp',
+    './requestAnimationFrame',
+    './DeveloperError',
+    './DoublyLinkedList'
+], function(
+    defined,
+    defineProperties,
+    getTimestamp,
+    requestAnimationFrame,
+    DeveloperError,
+    DoublyLinkedList) {
+    'use strict';
+
+    function prune(cache) {
+        var currentTime = getTimestamp();
+        var pruneAfter = currentTime - cache._expiration;
+
+        var list = cache._list;
+        var node = list.tail;
+        var index = list.length;
+        while (defined(node) && node.item.timestamp < pruneAfter) {
+            node = node.previous;
+            index--;
+        }
+
+        if (node === list.tail) {
+            return;
+        }
+
+        if (!defined(node)) {
+            node = list.head;
+        } else {
+            node = node.next;
+        }
+
+        while (defined(node)) {
+            delete cache._hash[node.item.key];
+            node = node.next;
+        }
+
+        list.removeAfter(index);
+    }
+
+    function checkExpiration(cache) {
+        function loop() {
+            if (!cache._hasExpiration || cache.length === 0) {
+                cache._expirationLoopRunning = false;
+                return;
+            }
+
+            cache._expirationLoopRunning = true;
+
+            prune(cache);
+
+            if (cache.length > 0) {
+                requestAnimationFrame(loop);
```

Comment:
I'm not sure that it's good practice to call `requestAnimationFrame` here or anywhere besides the main render loop in `CesiumWidget`. I don't think it will break anything, but it may request a render that isn't desired.

At the same time, I'm not sure what the alternative would be. Is there any way for the lru cache to be owned by something with access to frame state?

Reply:
I originally had `Scene.update` tell `sampleTerrain` to call `cache.checkExpiration`, but that results in that function being called every frame.  This way, LRUCache can manage it's own expiration and doesn't execute the loop when there's nothing in the cache.

Reply:
I agree that `LRUCache` is cleaner when it manages itself, I'm just worried about the calls to `requestAnimationFrame`.

Would it be not completely terrible to call `prune` after each `set` call? Or maybe use `setTimeout` if it doesn't need to prune every frame?

Reply:
I prefer what @hpinkos has done here.  It will not trigger a Cesium render, it will simply let LRUcache does it's processing all at once on the next browser (not Cesium) render

Reply:
My concern is that it will trigger a browser render unnecessarily. Or maybe it won't... maybe that's going to happen every frame regardless. But I still think `requestAnimationFrame` really only belongs inside the main render loop since that's where we "ask" for a frame. But if this is the cleanest way to get the job done, then I suppose it's ok.

Reply:
Two week later update. I read up on `requestAnimationFrame` more and it seems that it is purely callback aggregator and it won't force a render like I thought might be happening. Sorry if this is already common knowledge...

I still think its nice that `requestAnimationFrame` is called in just one place, but technically it should be fine to call in multiple.

Reply:
For the render loop, I agree incompletely, there should only be one `requestAnimationFrame` call.  For anything we want to have happen independent of the render loop, multiple `requestAnimationFrame` make total sense.  I think we need to continue to decouple things from our actual webgl render loop and move them one level up into `requestAnimationFrame` .